### DISCUSSION TITLE
What are the best practices for migrating from apisix 2 to apisix 3

### DISCUSSION THREAD

**Discussion by diaojibo** [apache/apisix#9207] [Q&A]
I want to upgrade my cluster's apisix version from 2 to 3, but I'm not sure what the best approach is. It seems infeasible to directly migrate the contents of a specific etcd key prefix to a new key, right? Can I only migrate the previous routes and consumers one by one through the admin API? Is there a better way?

**Comment by xgenvn** [comment]
I think you will need to go through this document: https://apisix.apache.org/docs/apisix/next/upgrade-guide-from-2.15.x-to-3.0.0/

So there are only some configurations to be relocated afaik. And some plugins will require updating the config.
I'm still checking the feasibility of the migration task, so my checklist is similar to below:

1. Get the list of current in-use plugins (or export all routes, then regex the plugins name)
2. Backup current routes
3. Setup new APISIX instance (v3)
4. Restore current routes
5. Check the configuration for related plugins
6. Run tests

If you have ingress or LB, it's also possible to:
1. Keep v2 running
2. Migrate route by route to v3 and make sure the route works
3. Migrate the LB rule to point to v3
4. Remove v2 when everything is working

In any case, adopt v3 as a regional proxy, and then move to global later.
